### PR TITLE
Intel GPU initial support

### PR DESCRIPTION
**Pull request by kurapov-peter**
This PR introduces `DeviceKernel`s and generalizes the GPU execution path. It also provides some minor fixes for code generation so that simple queries can now run on Intel GPU.

### CODE REVIEW

**Comment by kurapov-peter** [comment]
Tests pass for me locally for both CPU and CUDA builds.

File: QueryEngine/GpuMemUtils.cpp [hunk 1]
```diff
@@ -24,6 +24,8 @@
 
 #include <vector>
 
+using DevicePtr = int8_t*;
```
Comment (alexbaden) [inline QueryEngine/GpuMemUtils.cpp:27]: I'm surprised this still works with cuda enabled. Why do we not need an `#ifdef HAVE_CUDA` here similar to the `CUDevicePtr` definition?

File: QueryEngine/GpuMemUtils.h [hunk 2]
```diff
@@ -45,10 +45,10 @@ void copy_to_nvidia_gpu(Data_Namespace::DataMgr* data_mgr,
                         const int device_id);
 
 struct GpuGroupByBuffers {
-  CUdeviceptr ptrs;  // ptrs for individual outputs
-  CUdeviceptr data;  // ptr to data allocation
+  int8_t* ptrs;  // ptrs for individual outputs
```
Comment (alexbaden) [inline QueryEngine/GpuMemUtils.h:48]: Same comment as above.

File: QueryEngine/QueryExecutionContext.cpp [hunk 3]
```diff
@@ -301,59 +295,29 @@ std::vector<int64_t*> QueryExecutionContext::launchGpuCode(
...
+    kernel->launch(grid_size_x,
+                   grid_size_y,
+                   grid_size_z,
+                   block_size_x,
+                   block_size_y,
+                   block_size_z,
+                   shared_memory_size,
+                   kernel_params);
 
-    if (hoist_literals) {
```
Comment (alexbaden) [inline QueryEngine/QueryExecutionContext.cpp:320]: I don't see this branch in the new Cuda device kernel.

File: QueryEngine/GpuMemUtils.cpp (hunk 1 above)
Comment (kurapov-peter) [reply QueryEngine/GpuMemUtils.cpp:27]: Yeah, I'm relying on pointers sizes to be the same which isn't a good assumption, I agree.

Comment (kurapov-peter) [reply QueryEngine/GpuMemUtils.cpp:27]: Decided to work on a proper solution separately (#505), leaving the current solution as is for now.

File: QueryEngine/QueryExecutionContext.cpp (hunk 3 above)
Comment (kurapov-peter) [reply QueryEngine/QueryExecutionContext.cpp:320]: Thanks, fixed!

**Comment by kurapov-peter** [comment]
I was unable to reproduce the failure, the test `./ArrowBasedExecuteTest --gtest_filter=Select.FilterAndSimpleAggregation --use-groupby-buffer-desc` works for me.

**Review by alexbaden — APPROVED** [review]
Assuming this passes (or mostly passes) with literal hoisting enabled, changes look correct now.

**Comment by kurapov-peter** [comment]
> Assuming this passes (or mostly passes) with literal hoisting enabled, changes look correct now.

Turns out our jit-engine branch fails with literal hoisting disabled :) The default *enabled* works fine in both cases.

**Pull request merged**